src/main/java/io/confluent/kafkarest/SimpleConsumerPool.java

```diff
@@ -34,37 +35,192 @@
+    private Time time;
+    private final TopicPartition topicPartition;
+
+    RecordsFetcher(TopicPartition topicPartition) {
+      this.ownerPool = SimpleConsumerPool.this;
+      this.time = SimpleConsumerPool.this.time;
+      this.maxPollTime = SimpleConsumerPool.this.maxPollTime;
+      this.topicPartition = topicPartition;
+      initialized = false;
+    }
+
+    // support for lazy initialization gives opportunity
+    // to add some kind of cache in future.
+    private Consumer<byte[], byte[]> initializeConsumer(long offset) {
+      if (!initialized) {
+        consumer = ownerPool.get(topicPartition, offset);
+        initialized = true;
+      }
+      return consumer;
+    }
+
+    public List<ConsumerRecord<byte[], byte[]>> poll(final long offset, final long count) {
+      if (count == 0) {
+        return Collections.emptyList();
+      }
+
+      List<ConsumerRecord<byte[], byte[]>> result = new ArrayList<>();
+
+      long endTime = time.milliseconds() + maxPollTime;
+      long pollTime;
+
+      Iterator<ConsumerRecord<byte[], byte[]>> it = null;
+      boolean enough = false;
+
+      while (!enough &&
+          (pollTime = endTime - time.milliseconds()) > 0) {
+
+        if (it == null || !it.hasNext()) {
+
+          if (!initialized) {
+            initializeConsumer(offset);
+          }
+
+          ConsumerRecords<byte[], byte[]> records = consumer
+              .poll(Math.max(0, pollTime));
+          if (records.isEmpty()) {
+            continue;
+          }
+          it = records.iterator();
+        }
+
+        while (it.hasNext()) {
+          ConsumerRecord<byte[], byte[]> record = it.next();
+          result.add(record);
+          if (result.size() == count) {
+            enough = true;
+            break;
+          }
+        }
+      }
+      return result;
+    }
+
+    public void close() throws Exception {
+      if (initialized) {
+        ownerPool.release
+            (consumer, topicPartition, consumer.position(topicPartition));
+      }
+    }
   }
 
-  synchronized public SimpleFetcher get(final String host, final int port) {
+
+  /**
+   * @return assigned Consumer that is ready to be used for polling records
+   */
+  synchronized public Consumer<byte[], byte[]> get(TopicPartition topicPartition, long startOffset) {
 
     final long expiration = time.milliseconds() + poolInstanceAvailabilityTimeoutMs;
 
     while (true) {
-      // If there is a SimpleConsumer available
-      if (availableConsumers.size() > 0) {
-        final String consumerId = availableConsumers.remove();
-        return new SimpleFetcher(simpleConsumers.get(consumerId), this);
+
+      // If there is a KafkaConsumer available
+      if (!topicPartitionOffsetConsumers.isEmpty()) {
+        Map<Long, List<Consumer<byte[], byte[]>>> consumersMap =
+          topicPartitionOffsetConsumers.get(topicPartition);
+
+        Consumer<byte[], byte[]> consumer;
+
+        if (consumersMap != null) {
```

Comment:
This gets rather complicated. It might be worth trying to break it down into a couple of helper methods. A bit of the logic also seems duplicated between the two (i.e. the parts labeled "get the longest waiting consumer".